### ISSUE TITLE
Apply directives do not work as expected in `svelte-scoped`

### ISSUE THREAD

**Issue by scheatkode** [unocss/unocss#2680]
### UnoCSS version

0.52.4

### Describe the bug

Custom styles defined in imported CSS files are not transformed when using apply directives. For instance, the following, when defined in `app.s?css`:

```css
div {
  --at-apply: text-black;
}
```

Renders to:

```css
div {
  --at-apply: text-black;
}
```

Instead of:

```css
div {
  --a-text-opacity: 1;
  color: rgba(0,0,0,var(--a-text-opacity));
}
```

---

Furthermore, `theme()` is not recognized, even inside components.

```svelte
<p>...</p>

<style>
  p {
    color: theme("colors.a_color.a_shade");
  }
</style>
```

Renders to:

```css
p.s-<hash> {
  color: theme("colors.black");
}
```

Instead of:

```css
p.s-<hash> {
  color: #000;
}
```

---

I'm reporting this as a bug since these break the expectations set by the `svelte-scoped` mode prior to the rewrite leading to `@unocss/svelte-scoped`. But don't get me wrong, I think `@unocss/svelte-scoped` is a major step forward. Thanks to all the maintainers involved !

### Reproduction

The above examples should be sufficient, but please do tell if a full reproduction is required.

### Validations

- [X] Read the [Contributing Guidelines](https://github.com/unocss/unocss/blob/main/CONTRIBUTING.md).
- [X] Check that there isn't [already an issue](https://github.com/unocss/unocss/issues) that reports the same bug to avoid creating a duplicate.
- [X] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/unocss/unocss/discussions) or join our [Discord Chat Server](https://chat.antfu.me).
- [X] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

**Comment by r-moore** [comment]
I have found that if you use `<style lang="postcss">` svelte-scoped does not parse any styles at all.
I use `lang="postcss` to prevent VSCode marking `@apply` as invalid syntax.

**Comment by jacob-8** [comment]
> break the expectations set by the `svelte-scoped` mode prior to the rewrite leading to `@unocss/svelte-scoped`

Thanks for reporting! 

New users would not see this as a bug since the docs say "Transformers and extractors are not supported due to the differences in normal UnoCSS global usage and Svelte Scoped usage" and there is no mention of handling non Svelte files nor the `theme()` or `@screen` features found in [@unocss/transformer-directives](https://unocss.dev/transformers/directives).

But yes, the previous `svelte-scoped` offered usage of [@unocss/transformer-directives](https://unocss.dev/transformers/directives) so the migration expectation is broken.

To give a little history, [@unocss/transformer-directives](https://unocss.dev/transformers/directives) never fully worked with `svelte-scoped` because it had no ability to write class selectors such that context dependent classes like `dark:text-white` would work. Those were broken, but now they work. See https://unocss.dev/integrations/svelte-scoped#apply-directives. So we took a step back in features but a step forward in that all `@apply` classes now work.

I'm sure there are probably others but you are the first to mention that you use `theme()`, so +1 for community desire to see that feature here. If you want to see `theme()` usage land soon you could start by adding a failing test in packages\svelte-scoped\src\preprocess\transformApply\index.test.ts and then compare the files in that folder with how `@unocss/transformer-directives` works to see if you can create a working PR. Otherwise I will look into it at some point.

As to non-Svelte CSS files, one option could be to expose a `cssFileTransformers` option from the plugin that would then work the same as the standard Vite plugin's `transformers` option works, except it would not be applied to Svelte files. I also see a lot of value in being able to aim Uno at CSS files. This would be a separate PR.

**Comment by jacob-8** [comment]
> I have found that if you use `<style lang="postcss">` svelte-scoped does not parse any styles at all.

@r-moore please create a repro and a separate issue. I'm not able to recreate your issue and from a simple test the attribute is not causing problems:

<img width="856" alt="image" src="https://github.com/unocss/unocss/assets/7559478/932bbec3-bebf-429b-8bfb-9248f5f9049a">

**Comment by scheatkode** [comment]
My apologies, I opened the issue in a bit of a hurry. You're right in saying that `theme()` and `@screen` are provided by `@unocss/transformer-directives`.

> New users would not see this as a bug since the docs say "Transformers and extractors are not supported due to the differences in normal UnoCSS global usage and Svelte Scoped usage" and there is no mention of handling non Svelte files nor the `theme()` or `@screen` features found in [@unocss/transformer-directives](https://unocss.dev/transformers/directives).
>
> But yes, the previous `svelte-scoped` offered usage of [@unocss/transformer-directives](https://unocss.dev/transformers/directives) so the migration expectation is broken.

As you said, this may be a step backward in features and new users would expect to have similar functionality to the mainline UnoCSS since `svelte-scoped` effectively replaces that.

> I'm sure there are probably others but you are the first to mention that you use `theme()`, so +1 for community desire to see that feature here. If you want to see `theme()` usage land soon you could start by adding a failing test in packages\svelte-scoped\src\preprocess\transformApply\index.test.ts and then compare the files in that folder with how `@unocss/transformer-directives` works to see if you can create a working PR. Otherwise I will look into it at some point.
> 
> As to non-Svelte CSS files, one option could be to expose a `cssFileTransformers` option from the plugin that would then work the same as the standard Vite plugin's `transformers` option works, except it would not be applied to Svelte files. I also see a lot of value in being able to aim Uno at CSS files. This would be a separate PR.


I'm not sure I currently have enough bandwidth to handle this task but let me see what I can do. `theme()` and `@screen` are nice-to-haves but can be worked around so I'll attempt to prioritize exposing a `cssFileTransformers` option, which should be sufficient for the common case and bring `@unocss/svelte-scoped` to *almost* feature-parity with the old `mode: "svelte-scoped"` implementation.

**Comment by jacob-8** [comment]
I added `theme()` support in #2689 as it's a pretty lightweight addition. @scheatkode can you peak at that and confirm it's what you need.

I really don't want to add `@screen` as it's a lot heavier. The original transformer-directives plugin is a lot more complex in that it's recursive and therefore has to use a lot more offsets to calculate replacements. If I'm reading it correctly this is to make `@screen` usage possible as I've been able to avoid that complexity with just doing apply and theme.

My initial thoughts are that `@screen` usage can stay in the realm of dedicated CSS stylesheets that are imported. If such exists, can you please give a firm example of why `md:` prefixes can't be used instead inside Svelte component style blocks. I don't use `@screen` and am not aware of any situation where I can't use the breakpoint prefixes. It's just for code organization I believe, as you did say it's a nice-to-have.

**Comment by scheatkode** [comment]
> I added `theme()` support in #2689 as it's a pretty lightweight addition. @scheatkode can you peak at that and confirm it's what you need.

Will do. Also good timing since I just opened #2690; can you take a look and see if I'm in the right direction ?

> I really don't want to add `@screen` as it's a lot heavier. The original transformer-directives plugin is a lot more complex in that it's recursive and therefore has to use a lot more offsets to calculate replacements. If I'm reading it correctly this is to make `@screen` usage possible as I've been able to avoid that complexity with just doing apply and theme.
> 
> My initial thoughts are that `@screen` usage can stay in the realm of dedicated CSS stylesheets that are imported. If such exists, can you please give a firm example of why `md:` prefixes can't be used instead inside Svelte component style blocks. I don't use `@screen` and am not aware of any situation where I can't use the breakpoint prefixes. It's just for code organization I believe, as you did say it's a nice-to-have.

I agree, seeing the code first hand, transformers are quite complex. I also agree that `@unocss/svelte-scoped` and the transformers should not step on each others' toes, therefore I'm fine with keeping transformers in the realm of imported CSS to avoid the added complexity and maintenance burden. That should be more than sufficient for the common case as well as mine. Anything else can be easily worked around.

**Comment by jacob-8** [comment]
Good, this will be closed by #2689 and #2690.

**Comment by jacob-8** [comment]
Both https://github.com/unocss/unocss/pull/2689 and https://github.com/unocss/unocss/pull/2690 have been merged.
